### PR TITLE
Add endpoint for storing user email + add logging for query-results

### LINKED ISSUE #11: Send received data to an email
* Add an endpoint to set up the email address (PUT/email_address). Store it in datastore (https://cloud.google.com/datastore/docs).
* Modify endpoint developed in #5 to send received data to the specify email address.

### PR DESCRIPTION
**Pull request by niedziol**
Closes #11

### CODE REVIEW

File: webhook/main.py [hunk 1]
```diff
@@ -1,39 +1,46 @@
...
+
+
+@ndb.transactional
+def update_email(content: EmailWrapper):
+    with client.context():
+        q = UserEmail.query(api_key=content.api_key)
+        q_result = list(q.fetch())
+        if len(q_result) > 1:
+            raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
+                                detail="There exist multiple rows with same api key")
```
Comment (mgmacias95) [inline webhook/main.py:32]: Why check this on every request? This can be expensive and slow.

File: webhook/main.py [hunk 2]
```diff
@@ -1,39 +1,46 @@
...
+
+        obj = None
+        if len(q_result) == 0:
+            obj = UserEmail(api_key=content.api_key, email=content.email)
+        else:
+            obj = q_result[0]
```
Comment (mgmacias95) [inline webhook/main.py:38]: You're not updating email adreess, aren't you?

File: webhook/tests/test_main.py [hunk 3]
```diff
@@ -140,3 +141,22 @@ def test_query_results_missing_field_in_data():
             })
 
         assert r.status_code == 422
+
+
+def test_email_address_new(monkeypatch):
+    data = {
+        "api_key": "abcabc",
+        "email": "some@email.example"
```
Comment (mgmacias95) [inline webhook/tests/test_main.py:149]: Add a test with a wrong email and wrong API key too.

File: webhook/main.py (hunk 1 above)
Comment (niedziol) [reply webhook/main.py:32]: Well there should never occur situation when q_result has more then 1 row and I need to check if given API key exists in datastore so how can I do it better?

Comment (mgmacias95) [reply webhook/main.py:32]: Yes, you can do the following
```python
q = UserEmail.get_by_id(apikey)
```

Assuming you're using the API key as ID in datastore. Using email would also be valid.

Comment (niedziol) [reply webhook/main.py]: Done

File: webhook/main.py (hunk 2 above)
Comment (niedziol) [reply webhook/main.py]: Done

File: webhook/tests/test_email_address.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+    from main import app
+
+    client = TestClient(app)
+
+    r = client.post(
+        '/email-address/',
+        json=data
+    )
+    print(r.text)
```
Comment (mgmacias95) [inline webhook/tests/test_email_address.py:37]:
```suggestion
```

File: webhook/models.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    def api_key_validator(cls, v):
+        if not v.isalnum():
+            raise ValueError("API key must be alphanumeric!")
+        return v
+
+    @validator("email")
+    def email_validator(cls, v):
+        regex = "^[a-z0-9]+[\._]?[a-z0-9]+[@]\w+[.]\w+$"
+
+        if not re.search(regex, v):
```
Comment (mgmacias95) [inline webhook/models.py:43]: You can also use `re.match` (Read: https://docs.python.org/3/library/re.html#search-vs-match). I think `match` is more efficient.

File: webhook/tests/test_email_address.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+from models import UserEmail
+
+import pytest
+
+client = TestClient(app)
+
+URL = "/email-address/"
+
+
+@pytest.fixture(autouse=True)
```
Comment (mgmacias95) [inline webhook/tests/test_email_address.py:14]: I love this!

File: webhook/tests/test_models.py [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+def check_value_error(api_key="normalkey123123", email="some@email.address"):
+    try:
+        EmailWrapper(api_key=api_key, email=email)
+    except ValueError:
+        return True
+    except:
+        return False
+
+def test_email_wrapper_bad_api_key():
```
Comment (mgmacias95) [inline webhook/tests/test_models.py:21]: Check `pytest.mark.parametrize` (https://docs.pytest.org/en/stable/parametrize.html#pytest-mark-parametrize) to parametrize test functions

File: webhook/models.py (hunk 5 above)
Comment (niedziol) [reply webhook/models.py:43]: "Regular expressions beginning with '^' can be used with search() to restrict the match at the beginning of the string:"
But match is probably more readable, so I'll change that.

File: webhook/main.py [hunk 8]
```diff
@@ -1,39 +1,65 @@
...
+
+    email = None
+    with client.context():
+        email = ndb.Key("UserEmail", request.api_key).get()
 
+    if email is None:
+        raise HTTPException(
+            status_code=status.HTTP_403_FORBIDDEN, detail="API Key is not valid")
+    
+    log.info("Send an email to %s.\nBody:\n%s".format(email, json.dumps(jsonable_encoder(request.data))))
```
Comment (karpp) [inline webhook/main.py:39]: I think here must be one of `"%s ... %s" % (a, b)` or `"{} ... {}".format(a, b)`

File: webhook/main.py [hunk 9]
```diff
@@ -1,39 +1,65 @@
...
 
-class VTData(BaseModel):
-    attributes: Dict
-    id: str
-    links: Dict
-    type: VTType
 
+@app.post("/email-address/")
+async def set_email(response: Response, content: APIKeyEmail):
+    # TODO Check why documentation is not generating correct response code_status.
```
Comment (karpp) [inline webhook/main.py:44]: Do you still need this TODO?

File: webhook/tests/test_models.py [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+
+
+@pytest.mark.parametrize("email", ["without.domain", "mail@withoutsufix", "mail@withoutsufix",
+                                   "without.at.com", "strange@\x0char.com"])
+def test_email_wrapper_bad_email(email):
+    try:
+        APIKeyEmail(api_key="abc123", email=email)
+    except ValueError:
+        pass
+
```
Comment (karpp) [inline webhook/tests/test_models.py:22]: Add tests for correct emails too

File: webhook/tests/test_query_results.py [hunk 11]
```diff
@@ -83,29 +80,26 @@ def test_query_results_links_and_meta():
     }
 
     r = client.post(
-        '/query-results/',
+        URL,
         headers=correct_headers,
-        json={"data": [], "links": links, "meta": meta})
+        json={"api_key": API_KEY, "data": [], "links": links, "meta": meta})
 
     assert r.status_code == 200
```
Comment (karpp) [inline webhook/tests/test_query_results.py:87]: You can also check that result was logged

File: webhook/tests/test_models.py [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+        pass
+ 
+
+
+@pytest.mark.parametrize("email", ["without.domain", "mail@withoutsufix", "mail@withoutsufix",
+                                   "without.at.com", "strange@\x0char.com"])
+def test_email_wrapper_bad_email(email):
+    try:
+        APIKeyEmail(api_key="abc123", email=email)
+    except ValueError:
```
Comment (karpp) [inline webhook/tests/test_models.py:20]: What happens if it doesn't throw an exception?

File: webhook/main.py (hunk 9 above)
Comment (niedziol) [reply webhook/main.py:44]: Yeah, I have no idea how to make docs recognise that there can be code 201. It's seing only 200 and 422 for now.

File: webhook/tests/test_models.py (hunk 10 above)
Comment (niedziol) [reply webhook/tests/test_models.py:33]: Done

File: webhook/tests/test_models.py (hunk 12 above)
Comment (niedziol) [reply webhook/tests/test_models.py:31]: Fixed, now it'll return if it's ValueError and break if function doesn't raise anything.

File: webhook/main.py (hunk 8 above)
Comment (niedziol) [reply webhook/main.py]: Fixed

File: webhook/tests/test_models.py (hunk 10 above)
Comment (karpp) [reply webhook/tests/test_models.py:33]: I found cases that don't work
"with.two.dots@address.com", "with-dash@gmail.com"

Comment (niedziol) [reply webhook/tests/test_models.py:34]: Fixed

File: webhook/tests/test_query_results.py (hunk 11 above)
Comment (niedziol) [reply webhook/tests/test_query_results.py:87]: I think that we can trust that python logger is working.

**Pull request merged**